src/org/pentaho/reporting/platform/plugin/output/CachingPageableHTMLOutput.java

```diff
@@ -71,6 +76,12 @@
   @Override
   public int paginate( final MasterReport report, final int yieldRate )
     throws ReportProcessingException, IOException, ContentIOException {
+    final ExtendedConfiguration config = ClassicEngineBoot.getInstance().getExtendedConfig();
+    final boolean firstPageMode =
+      config.getBoolProperty( "org.pentaho.reporting.platform.plugin.output.FirstPageMode" );
+    if ( firstPageMode ) {
```

Comment:
No, dont do that. 

Change the report-viewer to not ask for paginated results (renderMode=XML vs renderMode=PARAMETER). When in async-mode, the report-viewer should never send parameter-xml requests with the renderMode=XML mode. Instead, compute and return the total number of pages as part of the status request, and then update the paginationControl in the report-viewer with those values.

This then renders your change here unnecessary as "paginate" will never be called on any output-handler.